rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,565 @@
+---
+authors: Noah Stride (noah.stride@goteleport.com)
+state: draft
+---
+
+# RFD 00162 - Improving the on-prem fleet Bot management experience
+
+## Required approvers
+
+* Engineering: @zmb3
+* Product: @klizhentas || @xinding33
+* Security: @reedloden || @jentfoo
+
+## What
+
+Terminology:
+
+- Bot: An identity within Teleport intended for use by machines as opposed to
+  humans. Many individual machines may act as this shared identity.
+- `tbot`: The Teleport binary that acts as aBot and generates credentials for
+  consumption by client applications.
+- Bot instance: A single instance of `tbot` running on a host.
+
+This RFD proposes improvements to the management of fleets of Machine ID Bots.
+These improvements are mostly targetted at on-prem deployments, where the
+delegated join methods are not available.
+
+The improvements will focus on three points:
+
+- To allow multiple Bot instances to be associated with a single Bot when using
+  the `token` join method.
+- To allow multiple Bot instances to be joined using a single join token when
+  using the `token` join method.
+- Providing a way to track and monitor Bot instances.
+
+## Why
+
+Whilst deploying a large fleet of Bots is fairly trivial when using the
+delegated join methods, the experience when managing a fleet of bot hosts
+in-prem is more challenging. 
+
+The following burdens currently exist:
+
+- When using the `token` join method, a Bot must be created for each host. This
+  means that the privileges of many distinct Bots need to be synchronised where
+  those hosts are performing the same function.
+- When using the `token` join method, a token can only be used once. This means
+  creating hundreds of join tokens and managing securely distributing these to
+  hosts.
+- When managing a large fleet of `tbot` deployments, there is no way to
+  track these within Teleport. This makes it more difficult to identify hosts
+  which may need updating.
+
+As we look to onboard more Enterprise customers to Machine ID, the pains of
+the on-prem experience have become more apparent. Enterprise customers are
+more likely to have on-prem deployments and these are likely to be larger in
+scale.
+
+## Details
+
+### Current State
+
+Currently, the `token` join method introduces a generation counter as a label
+on the Bot user. This counter is contained within the Bot certificate and on
+each renewal, this counter is incremented. When the counter within the certificate
+de-synchronises with the counter on the user, the Bot is locked out as a security
+measure.
+
+The fact that this counter is stored within a label on the Bot user creates a 
+one-to-one binding between a single instance of `tbot` and a single Bot user.
+This is not the case when using the delegated join methods.
+ 
+### Persistent Bot Instance Identity
```

Comment:
After some experimentation, I think we'll actually need to do most of the more complex work involved in UUID verification even if we pick public key identifiers.

The problem, I think, is that `RegisterUsingTokenRequest` doesn't check the public key regardless, so clients could just lie about it. This means a malicious bot could potentially impersonate an existing instance. I don't think that's a hugely critical vulnerability but might help a malicious bot disguise itself.

I think we'll want to pass along an existing certificate to verify regardless... and we have to verify private key ownership, so that means we still need to return the new certs encrypted with the provided public key, after verifying it against auth's CA.

With that in mind, I don't think it's much of a stretch to include additional fields in the cert if we want.

It might honestly also be worthwhile to look at adding a grpc entrypoint to delegated joining. We'd then get a proper TLS session with the existing client cert which is IMO much less sketchy than manually verifying the certs and re-encrypting the payload from the existing endpoint.

Reply:
Yeah - that's a good point. Although, even if we pursue the simple model of encrypting the returned certs, for the purposes of the audit logs, a malicious actor can still pretend to be another bot instance. I thought about requiring a signature when making the initial request but I worry about how those can relatively easily be re-sent unless the signed value is some sort of challenge - perhaps use a timestamp ???

As you say, a gRPC entrypoint would resolve most of these concerns and would allow for any challenge and response flows we wanted. It'd also be an easy way to flip between the old and new behaviour in regards to BotInstances. I think the only challenge worth bearing in mind with a gRPC entrypoint is how an unauthenticated client would reach the auth server, today, the IAM and Azure join methods are terminated at the Proxy and the gRPC requests are re-made, but this would conceal the TLS certificate from the auth server. Perhaps that's not really a problem and unauthenticated requests could be made to the proxy directly, and when you have an existing cert, you can tunnel through and hit the auth server directly.

I like the idea generally and moving away from HTTP and inline with other efforts - my only major concern with moving this to gRPC would be avoiding scope creep. There's a lot in the RegisterUsingToken flow that people will want to refactor and change, and we'll need to resist that if we want to avoid this becoming a rather large task. Perhaps we set out a bounds and say for the V1 of the gRPC method, we'll mimic the HTTP fields precisely.

Reply:
>  worry about how those can relatively easily be re-sent unless the signed value is some sort of challenge - perhaps use a timestamp ???

Yep, I tried to game out what a proper challenge would look like and was not liking the amount of effort or complexity it would add given it still feels like a hack vs a gRPC registration entrypoint.

> I like the idea generally and moving away from HTTP and inline with other efforts - my only major concern with moving this to gRPC would be avoiding scope creep

:+1: I'll pivot my investigation today to gauge how complex a gRPC refresh would be.

To cut scope initially I'd potentially be okay with a gRPC endpoint that is _only_ accessible by already authenticated users. Clients would onboard through HTTP, but renew via gRPC. I'll experiment and see what makes sense, if anything.

Reply:
> Yep, I tried to game out what a proper challenge would look like and was not liking the amount of effort or complexity it would add given it still feels like a hack vs a gRPC registration entrypoint.

Yeah the only thing that comes to mind is signing a timestamp and having the auth verify that it falls within a 5 minute window - but that's not fantastic for a bunch of reasons.

> 👍 I'll pivot my investigation today to gauge how complex a gRPC refresh would be.

Sounds good - gRPC gives us the option to do call and responses which is great. The TPM joining I did fairly recently is a nice-ish example of how to achieve this and does show how to use the existing JoinUsingTokenRequest proto within a challenge-response flow.

Reply:
Alright, I'll be trying to merge our discussions back into the RFD today. My summary so far is:
* We'd ideally like to use UUIDs (or similar) as the primary bot identifier, to allow (among other things) key rotation; I'll rewrite this section to list that as the preferred method and public keys as the alternate
* I'll also add a short section on trust of this new identifier to help outline why I think it's important to verify it during renewal.